packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -577,7 +579,9 @@ class SliverConstraints extends Constraints {
   int get hashCode => Object.hash(
     axisDirection,
     growthDirection,
+    userScrollDirection,
     scrollOffset,
+    precedingScrollExtent,
```

Comment:
Could you please add a test that verifies that hashCode and == consider these two properties so we don't break this again in the future? Thanks!

Reply:
I can add a test, but such a test wouldn't have helped prevent what broke this. When this field was added back in #24551, it wasn't added to hashCode and == in the first place. If there'd been a test of this form, then the test wouldn't have broken either.

I don't know of a reasonable way of writing a test that hashCode and == are comprehensive, in a way that would successfully break when someone adds a new field and forgets to update those methods. (It could be done by a test that examines the source code, either with `package:analyzer` or by hand, if we were determined to do so.) I don't see any existing examples in the codebase that do that — there are some tests that `copyWith` preserves `hashCode`  and `==`, and a few tests that specific fields get used in `hashCode` (like in `test/material/input_decorator_test.dart`), but none that would notice a new field.